### ISSUE TITLE
New infrastructure for DOAJ

### ISSUE THREAD

**Issue by emanuil-tolev** [DOAJ/doaj#605]
HTTPS everywhere, new infrastructure.

Current temporary test server at https://doaj.cottagelabs.com . Will probably establish test_doaj.cottagelabs.com and staging_doaj.cottagelabs.com after PVOD release in April.

@emanuil-tolev to answer this: What happens to referrer information for Google Analytics when somebody comes to https://doaj.org from a non-secure site? Will DOAJ use referral data? (DOAJ team uses it now in GA reports)

**Comment by markmacgillivray** [comment]
It is unclear what sort of answer you expect. If someone goes to a wrong
url and is redirected, then the referrer data may be dropped. So if people
are pointing to http instead of https then their redirect may be
incorrectly logged. But, over time people should start using the https
addresses because they are the only ones that the new service will display
- so for example when they copy and paste them they will have no choice.

So there may be a while where https redirects cause direct traffic to
appear higher than it is. But once people update their links this should
fade away.

This may be an inconvenience, but it is worth making the service more
secure.

Mark

**Comment by emanuil-tolev** [comment]
> If someone goes to a wrong
> url and is redirected, then the referrer data may be dropped.

301-s very very often keep the referrer.

DOAJ are currently using GA to analyse where traffic comes from, i.e. which sites are sending users to doaj.org . When you go from a HTTPS to a non-secure site the browser wipes the referrer. I thought it was the other way around as well. So I thought that putting HTTPS everywhere would remove a lot of the information the DOAJ team are using in GA. But this hypothesis is likely wrong - the fact it works like that from secure to insecure doesn't mean the opposite's true :). So the above is just a pointer for me to double check this detail.

**Comment by markmacgillivray** [comment]
Indeed. So I don't think it's an issue. Is there anything else you want me
to do with this?

**Comment by emanuil-tolev** [comment]
> Is there anything else you want me
> to do with this?

nope, I was just leaving the note for myself (and keeping this issue to track other issues with the new infrastructure), thanks

**Comment by markmacgillivray** [comment]
Ok just checking, it's assigned to me so wanted to know if I should do
something.

**Comment by emanuil-tolev** [comment]
Another issue I thought of here: when there is a problem that requires looking at the logs to figure out what happened, having the app deployed over X number of servers will make it difficult to find the right log file. Possible solutions are some kind of concatenation, sending to loggly or another platform like it, sending to our own instance of ELK (ES, Logstash, Kibana) for 14-28 days.

We have had to use the log rather than reproduce on a test server only 5-6 times, so perhaps this is not vitally urgent, but we should consider what to do about it. At the end of the day we can of course simply look in all app servers when bad things happen.

**Comment by markmacgillivray** [comment]
Ready to go

**Comment by markmacgillivray** [comment]
Note. Need to check where stuff gets written to disk. Should all write to storage service directory. Then crons should run on a cron machine.

**Comment by markmacgillivray** [comment]
May write to temp directory to check performance, then cron onto storage service directory on each app machine for checking. Then run crons of the doaj on a cron machine.

Files can be written to storage of local machine if they don't need to be synced across apps machines, but cannot write into the directory of the running app - all code for the apps gets put there by replication from the gateway machine, so writing anything into the running app folder risks it being wiped on the next replication. Apps must function as if everything is an attached service, including storage directory.

**Comment by emanuil-tolev** [comment]
May need to collab on a plan for the rollout - on Monday or today? (if rollout's still scheduled for Mon 13 Apr - wasn't at last meeting, sorry)

Sounds like this change (the storage) needs to be done before rollout. I'll flag for @richard-jones today at 3pm tech meeting (feel free to come if you're free, could be a good time to talk about anything else outstanding for the live deploy).

**Comment by richard-jones** [comment]
Leave this open until production deployment, then can close

**Comment by emanuil-tolev** [comment]
When turning on readonly mode, turn it on for the UI first at 8:21pm. Then let the scripts digest whatever has been recently uploaded at 8:30pm.Then turn on readonly mode for the scripts too. Now the whole site is readonly and the migration can start.

**Comment by emanuil-tolev** [comment]
Final deploy plan:

So:
1. perform code release from develop to production (5-10 min) without deploying it, obviously
2. turn on readonly mode in the evening (8:21pm UK time)
    
When turning on readonly mode, turn it on for the UI first at 8:21pm. Then let the scripts digest whatever has been recently uploaded at 8:30pm.Then turn on readonly mode for the scripts too. Now the whole site is readonly and the migration can start.

3. stop all regular jobs and the backups
4. initiate a data backup (45min)
5. download the backup to new infrastructure (15 min)
6. decompress the backup (15 min)
7. restore the backup to the new infrastructure (2h 15min)
8. delete /toc
9. in one single command, let migration scripts run overnight (3h+?) and let the index be optimized by Elasticsearch (reduce size from 33GiB to 15GiB, takes 4-6 more hours!) - this consolidates it after the initial bulk restore and the migrations and we proceed with a relatively clean/fast/small slate.
10. turn readonly mode off in the new site
11. switch doaj.org DNS to the dedicated doaj gateway - around 9:30am UK time Wednesday.

Everything will have a separate log of course, and the next command will only proceed if the previous was successful.

**Comment by emanuil-tolev** [comment]
tmp command here

```stdbuf -o0 -e0 /opt/sysadmin/backup/elasticsearch/bulk_restore.py -es "http://gateway:9200" --no-mapping -i doaj -b 200 -s 0.075 doaj.data > doaj_restore.log 2>&1 && curl -X DELETE http://gateway:9200/doaj/toc && . /home/cloo/repl/apps/doaj/bin/activate && DOAJENV=production stdbuf -o0 -e0 /home/cloo/repl/apps/doaj/bin/python /home/cloo/repl/apps/doaj/src/doaj/portality/upgrade.py -u /home/cloo/repl/apps/doaj/src/doaj/portality/migrate/pvod/pvod.json > ~/doaj_pvod_upgrade.log && curl -XPOST "http://gateway:9200/doaj/_optimize?max_num_segments=1"```

**Comment by emanuil-tolev** [comment]
> switch doaj.org DNS to the dedicated doaj gateway - around 9:30am UK time Wednesday.

everything went well, switching DNS in half an hour

**Comment by emanuil-tolev** [comment]
DNS switched. There were some problems with the blog news on the front page, generating a csv and generating a sitemap, all resolved now. All cron and anacron tasks are running.
